### DISCUSSION TITLE
TypeError: __call__() missing 1 required positional argument: 'send'

### DISCUSSION THREAD

**Discussion by Zhaoxun** [django/channels#1736] [Help!] [answered]
relating code in consumers.py:
```
from channels.generic.websocket import AsyncWebsocketConsumer
import json

class Trade(AsyncWebsocketConsumer):

    async def my_send(self, data):
        if isinstance(data, dict):
             d = json.dumps(data)
        else:
            d = str(data)
        await self.send(d)
```

OS =Windows 7-64bit
Python = 3.7.9

pip-freeze:
```
daphne==3.0.2
Django==3.2.6
docopt==0.6.2
et-xmlfile==1.0.1
future==0.18.2
hyperlink==21.0.0
idna==2.10
importlib-metadata==3.7.2
incremental==21.3.0
numpy==1.20.1
openpyxl==3.0.7
pandas==1.2.3
pefile==2019.4.18
pipdeptree==2.0.0
pipreqs==0.4.10
pyasn1==0.4.8
pyasn1-modules==0.2.8
pycparser==2.20
pyinstaller==4.2
pyinstaller-hooks-contrib==2021.1
pyOpenSSL==20.0.1
PyQt5==5.13.0
PyQt5-sip==12.8.1
pyqt5-tools==5.13.0.1.5
python-dateutil==2.8.1
python-dotenv==0.15.0
pytz==2021.1
pywin32-ctypes==0.2.0
requests==2.25.1
service-identity==21.1.0
six==1.15.0
sqlparse==0.4.1
Twisted==21.7.0
twisted-iocpsupport==1.0.1
txaio==21.2.1
typing-extensions==3.7.4.3
urllib3==1.26.4
websocket-client==1.1.1
websockets==8.1
xlrd==1.2.0
XlsxWriter==1.3.7
xlwt==1.3.0
yarg==0.1.9
zipp==3.4.1
zope.interface==5.4.0
```

full Traceback:

daphne.server - ERROR - Exception inside application: __call__() missing 1 required positional argument: 'send'
Traceback (most recent call last):
  File "C:\Python37\Lib\site-packages\channels\staticfiles.py", line 44, in __call__
    return await self.application(scope, receive, send)
  File "C:\Python37\Lib\site-packages\channels\routing.py", line 71, in __call__
    return await application(scope, receive, send)
  File "C:\Python37\Lib\site-packages\channels\sessions.py", line 47, in __call__
    return await self.inner(dict(scope, cookies=cookies), receive, send)
  File "C:\Python37\Lib\site-packages\channels\sessions.py", line 263, in __call__
    return await self.inner(wrapper.scope, receive, wrapper.send)
  File "C:\Python37\Lib\site-packages\channels\auth.py", line 185, in __call__
    return await super().__call__(scope, receive, send)
  File "C:\Python37\Lib\site-packages\channels\middleware.py", line 26, in __call__
    return await self.inner(scope, receive, send)
  File "C:\Python37\Lib\site-packages\channels\routing.py", line 160, in __call__
    send,
  File "C:\Python37\Lib\site-packages\asgiref\compatibility.py", line 35, in new_application
    return await instance(receive, send)
TypeError: __call__() missing 1 required positional argument: 'send'

**Comment by carltongibson** [comment]
Likely you missed the `as_asgi()` in your `routing.py`

**Reply by vitormhenrique** [reply]
how does the routing needs to be changed (from 2.4) ?

Is there an example how the as_asgi needs to be added?

**Reply by Zhaoxun** [reply]
```
from django.urls import path
from channels.routing import ProtocolTypeRouter, URLRouter
from channels.auth import AuthMiddlewareStack
from .consumers import Trade
from configobj import ConfigObj
from Profile import profile
conf = ConfigObj(profile.CONFIG_INI_URL,encoding="utf-8")
application = ProtocolTypeRouter({

    # WebSocket chat handler
    "websocket": AuthMiddlewareStack(
        URLRouter([
            path(conf["intranet"]["address"].split("/")[3]+"/", Trade)
        ])
    ),
})
```
Could you specify where in this  "routing.py" is to change?

**Reply by Zhaoxun** [reply]
Here are the lines involving asgi in "settings.py":
```
ASGI_APPLICATION = "my_project.routing.application"
''' CHANNEL_LAYERS = {
    "default": {
        "BACKEND": "channels_redis.core.RedisChannelLayer",
        "CONFIG": {
            "hosts": [("localhost", 6379)],
        },
    },
} '''
```
And here is the entire "asgi.py":
```
import os
import django
from channels.routing import get_default_application

os.environ.setdefault("DJANGO_SETTINGS_MODULE", "my_project.settings")
django.setup()
application = get_default_application()
```

**Reply by Zhaoxun** [reply]
And the running command is:
`python manage.py runserver 0.0.0.0:[PORTNUM]`

**Reply by nstadelmaier-dev** [reply]
> ```
> from django.urls import path
> from channels.routing import ProtocolTypeRouter, URLRouter
> from channels.auth import AuthMiddlewareStack
> from .consumers import Trade
> from configobj import ConfigObj
> from Profile import profile
> conf = ConfigObj(profile.CONFIG_INI_URL,encoding="utf-8")
> application = ProtocolTypeRouter({
> 
>     # WebSocket chat handler
>     "websocket": AuthMiddlewareStack(
>         URLRouter([
>             path(conf["intranet"]["address"].split("/")[3]+"/", Trade)
>         ])
>     ),
> })
> ```
> 
> Could you specify where in this "routing.py" is to change?

Try changing ` path(conf["intranet"]["address"].split("/")[3]+"/", Trade)` to ` path(conf["intranet"]["address"].split("/")[3]+"/", Trade.as_asgi())`
Note the `as_asgi()` call.
When using Channels 3, one has to call `as_asgi` on the consumer. I think that is what @carltongibson meant.

**Comment by praveenyadav1602** [comment]
my consumer.py 

```
from channels.generic.websocket import WebsocketConsumer
from asgiref.sync import async_to_sync
import json

class TestConsumer(WebsocketConsumer):

    def connect(self):
        self.room_name = 'test_consumer'
        self.room_group_name = 'test_consumer_group'
        async_to_sync(self.channel_layer.group_add)(self.room_name,self.room_group_name)
        self.accept()
        self.send(text_data=json.dump({'status':'connected praveen'}))

    def receive(self, text_data=None, bytes_data=None):
        pass

    def disconnect(self):
        pass      


```
my aggi.py

```

"""
ASGI config for websocket project.

It exposes the ASGI callable as a module-level variable named ``application``.

For more information on this file, see
https://docs.djangoproject.com/en/3.1/howto/deployment/asgi/
"""

import os

from django.core.asgi import get_asgi_application

from channels.routing import ProtocolTypeRouter, URLRouter
from django.urls import path
from mygameapp.comsumers import *
os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'websocket.settings')

application = get_asgi_application()
ws_patterns = [

path('ws/test/',TestConsumer)

]

application = ProtocolTypeRouter({

    'websocket' : URLRouter(ws_patterns)
})
```
i am also got same error i have not router.py, i wrote my ws pattern in asgi, how to solve

**Reply by shanda0409** [reply]
ws_patterns = [

path('ws/test/',TestConsumer.as_asgi())

]

will solve the issue

**Comment by Zhaoxun** [comment, marked as answer]
Try changing `path(conf["intranet"]["address"].split("/")[3]+"/", Trade)` 
to `path(conf["intranet"]["address"].split("/")[3]+"/", Trade.as_asgi())`
As nstadelmaier-dev suggested.
